src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,551 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+
+#include <ipc/topology.h>
+#include <sof/audio/tdfb/tdfb_comp.h>
+#include <sof/lib/alloc.h>
+#include <sof/math/iir_df2t.h>
+#include <sof/math/trig.h>
+#include <sof/math/sqrt.h>
+#include <user/eq.h>
+#include <stdint.h>
+
+/* Generic definitions */
+#define COEF_DEG_TO_RAD		Q_CONVERT_FLOAT(0.017453, 15)	/* Q1.15 */
+#define COEF_RAD_TO_DEG		Q_CONVERT_FLOAT(57.296, 9)	/* Q6.9 */
+#define PI_Q12			Q_CONVERT_FLOAT(3.1416, 12)
+#define PIMUL2_Q12		Q_CONVERT_FLOAT(6.2832, 12)
+#define PIDIV2_Q12		Q_CONVERT_FLOAT(1.5708, 12)
+
+/* Sound levels filtering related, these form a primitive voice activity
+ * detector. Sound levels below ambient estimate times threshold (kind of dB offset)
+ * are not scanned for sound direction.
+ */
+#define SLOW_LEVEL_SHIFT	11
+#define FAST_LEVEL_SHIFT	1
+#define POWER_THRESHOLD		Q_CONVERT_FLOAT(15.849, 10)	/* 12 dB, value is 10^(dB/10) */
+
+/* Iteration parameters, smaller step and larger iterations is more accurate but
+ * consumes more cycles.
+ */
+#define AZ_STEP			Q_CONVERT_FLOAT(0.6, 12)	/* PI/16 as Q4.12 */
+#define AZ_ITERATIONS		8				/* loops in min err search */
+#define SOURCE_DISTANCE		Q_CONVERT_FLOAT(3.0, 12)	/* source distance in m Q4.12 */
+
+/* Sound direction angle filtering */
+#define SLOW_AZ_C1		Q_CONVERT_FLOAT(0.02, 15)
+#define SLOW_AZ_C2		Q_CONVERT_FLOAT(0.98, 15)
+
+/* Threshold for notifying user space, no more often than every 100 ms */
+#define CONTROL_UPDATE_MIN_TIME	Q_CONVERT_FLOAT(100e-3, 16)
+
+/* Emphasis filter for sound direction */
+static const uint32_t iir_emphasis_48k[13] = {
+	0x00000001, 0x00000001, 0x00000000, 0x00000000,
+	0x00000000, 0x00000000, 0xd7b49168, 0x62c27f22,
+	0x19712b75, 0xcd1da916, 0x19712b75, 0x00000000,
+	0x00007fb2
+};
+
+static const uint32_t iir_emphasis_16k[13] = {
+	0x00000001, 0x00000001, 0x00000000, 0x00000000,
+	0x00000000, 0x00000000, 0xef646e59, 0x2cc88203,
+	0x0fb60984, 0xe093ecf7, 0x0fb60984, 0x00000000,
+	0x00007fb2
+};
+
+/* Function to copy data from component source to cross correlation buffer with emphasis
+ * filtering. Skip if not tracking.
+ */
+void tdfb_direction_copy_emphasis(struct tdfb_comp_data *cd, int ch_count, int *ch, int32_t x)
```

Comment:
can you document the variables here, passing and modifying ch is not clear

Reply:
done